### PR TITLE
Added 3rd type of slot search option which is optimized for API calls.

### PR DESCRIPTION
**Pull request by BlackRider97**
# Search option1:  Search by multiple PIN codes. 
 -  General use when you are in tier-III or town or village where single PIN code is sufficient
 - N PIN codes will be counted as N different API calls
 - **Granular but more API calls**

# Search option2:  Search by multiple district of a state
 -  General use when you are OK to take vaccine in anywhere in district. Usually tier-I cities.
 - N districts will be counted as N different API calls
- **Broader but less API calls**

# Search option3:  Search by multiple district of a state but filter by PIN codes
 -  General use when you are OK to take vaccine in anywhere in district. Usually tier-I cities.
 - Search#2 filter by specified PIN codes 
 - N districts will be counted as N different API calls
  - **Granular but less API calls too**

### CODE REVIEW

**Comment by bombardier-gif** [comment]
Thanks for the PR, but we making the UI too complex. 
When would the user use option 3 instead of option 1/2?

My opinion that our script is for people who are desperate for a vaccine. In a bunch of cities such as bangalore there are apartment wide drives coming anyway. So I will hold off on such complications for couple of days. 

Unless @BlackRider97 you have a strong opinion otherwise?

**Comment by BlackRider97** [comment]
To give my use case 

Jaipur-I and Jaipur-II district have few areas within main Jaipur city but there are few pincodes which are more 100 KM away which is not desired for who living in the main city. 

Same thing I have observed in Pune district where person do not want to go to remote areas but OK to take any where nearby PIN codes.

Option-2 is too broad where covering remote areas too 
Option-1 it too narrow and costly in terms of API calls

Option-3 is nothing but option-1 but better lesser API calls

**Comment by bombardier-gif** [comment]
Thanks for the context. Have we tested this PR?

**Comment by shailesh** [comment]
> Thanks for the PR, but we making the UI too complex.
> When would the user use option 3 instead of option 1/2?
> 
> My opinion that our script is for people who are desperate for a vaccine. In a bunch of cities such as bangalore there are apartment wide drives coming anyway. So I will hold off on such complications for couple of days.
> 
> Unless @BlackRider97 you have a strong opinion otherwise?

+1 I've the same openion.

**Comment by shailesh** [comment]
> To give my use case
> 
> Jaipur-I and Jaipur-II district have few areas within main Jaipur city but there are few pincodes which are more 100 KM away which is not desired for who living in the main city.
> 
> Same thing I have observed in Pune district where person do not want to go to remote areas but OK to take any where nearby PIN codes.
> 
> Option-2 is too broad where covering remote areas too
> Option-1 it too narrow and costly in terms of API calls
> 
> Option-3 is nothing but option-1 but lesser API calls

Had same thought earlier and had raised an issue but then we did not consider this because things would get more complicated and complex since motive for user is just to get a vaccine mentioned earlier by @bombardier-gif and then we closed such changes  

Anyway if testing is we can take this as a feature.

**Comment by Nakul93** [comment]
@BlackRider97 Its a good feature but sometimes we encounter some bugs after a merge, so I would suggest @bombardier-gif to create a separate branch for features and ppl can submit their feature PRs in that branch. So even if anyone wants to use it they can switch to branch and if nothing breaks then we can rebase it to main later

**Comment by BlackRider97** [comment]
Yes it is tested

**Comment by 6mxo89sqh** [comment]
@BlackRider97 Awesome feature! Same for few districts in Gujarat too. 

@Nakul93 creating a separate branch for features will be a headache for the maintainers to keep rebasing. Instead, using releases feature of github is a better alternative. You can use Github actions like [this one](https://github.com/release-drafter/release-drafter) to automate a lot of your release processes.

**Comment by Nakul93** [comment]
> @BlackRider97 Awesome feature! Same for few districts in Gujarat too.
> 
> @Nakul93 creating a separate branch for features will be a headache for the maintainers to keep rebasing. Instead, using releases feature of github is a better alternative. You can use Github actions like [this one](https://github.com/release-drafter/release-drafter) to automate a lot of your release processes. (Even the binaries viz. apk and exes can be moved to releases.)

Not if you are good at git, but yeah we could try out the release feature of GitHub.
Always pushing to main is generally not a good practice

**Comment by 6mxo89sqh** [comment]
> > @BlackRider97 Awesome feature! Same for few districts in Gujarat too.
> > @Nakul93 creating a separate branch for features will be a headache for the maintainers to keep rebasing. Instead, using releases feature of github is a better alternative. You can use Github actions like [this one](https://github.com/release-drafter/release-drafter) to automate a lot of your release processes. (Even the binaries viz. apk and exes can be moved to releases.)
> 
> Not if you are good at git, but yeah we could try out the release feature of GitHub.
> Always pushing to main is generally not a good practice

It's not being good with git. You will have to keep rebasing. I am pointing out that headache. 
Most open-source softwares, use same strategy. E.g. https://github.com/GoogleCloudPlatform/gsutil 

Industry is moving towards Trunk based development. So, it is actually it is good to push to trunk.
References:
https://cloud.google.com/architecture/devops/devops-tech-trunk-based-development
https://trunkbaseddevelopment.com/

**Comment by usthakur41** [comment]
> To give my use case
> 
> Jaipur-I and Jaipur-II district have few areas within main Jaipur city but there are few pincodes which are more 100 KM away which is not desired for who living in the main city.
> 
> Same thing I have observed in Pune district where person do not want to go to remote areas but OK to take any where nearby PIN codes.
> 
> Option-2 is too broad where covering remote areas too
> Option-1 it too narrow and costly in terms of API calls
> 
> Option-3 is nothing but option-1 but lesser API calls

+1 for this feature. I am from Pune and since there are soo many pin codes in the Pune district that are far away from Pune city, I tried to check availability for a few nearby pin codes 
 `411023,411037,411038,411001,411011,411058,411052,411004,411006,412115` but IP is getting blocked after 4-5min for 10sec interval due to too many API call.
This feature should come in handy in such cases.

**Comment by bombardier-gif** [comment]
@BlackRider97 Merging this in. Thanks for the contribution. Please watch out for new issues for a couple of days

File: src/utils.py [hunk 1]
```diff
@@ -571,7 +581,31 @@ def check_and_book(
...
+
+            pincode_filtered_options = []
+            for option in pincode_filtered_options: 
+                for location in pin_code_location_dtls:
+                    if int(location["pincode"]) in [option["pincode"] for option in options]:
+                        # ADD this filtered PIN code option
+                        pincode_filtered_options.append()
+                        for _ in range(2):
+                            beep(location["alert_freq"], 150)
+            options = pincode_filtered_options
```
Comment (6mxo89sqh) [inline src/utils.py:596]: `            pincode_filtered_options = []
            for option in options:
                for location in pin_code_location_dtls:
                    if int(location["pincode"]) == option["pincode"]:
                        # ADD this filtered PIN code option
                        pincode_filtered_options.append(option)
                        for _ in range(2):
                            beep(location["alert_freq"], 150)
            options = pincode_filtered_options
`

Comment (6mxo89sqh) [inline src/utils.py:596]:
```python
            pincode_filtered_options = []
            for option in options:
                for location in pin_code_location_dtls:
                    if int(location["pincode"]) == option["pincode"]:
                        # ADD this filtered PIN code option
                        pincode_filtered_options.append(option)
                        for _ in range(2):
                            beep(location["alert_freq"], 150)
            options = pincode_filtered_options
```

**Comment by 6mxo89sqh** [comment]
@BlackRider97 The code has a bug which shows no options despite options being available. I have commented for changes. Please check.

**Comment by BlackRider97** [comment]
@6mxo89sqh Raised PR to fix this #298

**Pull request merged**